[DOC] Add documentation page on denoising approaches (#823)
ME-ICA/tedana: docs/denoising.rst

##############################
Denoising Data with Components
##############################

Decomposition-based denoising methods like ``tedana`` will produce two important outputs:

> component time series and component classifications.

Reviewer: This would be perfect, yes!

``tedana`` uses independent component analysis (ICA) to decompose the data into components
which each have a spatial map of weights and time series that are assumed to reflect meaningful underlying signals.
It then classifies each component as "accepted" (BOLD-like) or "rejected" (non-BOLD-like).
The data are denoised by taking the time series of the rejected components and regressing them from the data.

``tedana`` uses a spatial ICA, which means the components are `statistically independent` across space, not time.
That means the time series from accepted and rejected components can share variance.
If a rejected component shares meaningful variance with an accepted component,
then regressing the rejected components' time series from your data may also remove meaningful signal
associated with the accepted component as well.
Depending on the application,
people may take different approaches on how to handle variance that is shared between accepted and rejected components.
These different approaches towards decomposition-based denoising methods are described here.

This page has three purposes:

1.  Describe different approaches to denoising using ICA components.

> 2.  Provide sample code using common code to perform each type of denoising.

Reviewer 2: ```suggestion
2.  Provide sample code that performs each type of denoising.
```

3.  Describe how to incorporate external regressors (e.g., motion parameters) into the denoising step.

.. admonition:: Which should you use?

  The decision about which denoising approach you should use depends on a number of factors.

  The first thing to know is that aggressive denoising, while appropriate for orthogonal time series,
  may remove signal associated with "good" time series that are independent to, but not orthogonal with,
  the "bad" time series being regressed out.

  As an alternative, users may want to try non-aggressive denoising or aggressive denoising combined with component orthogonalization.

  The main difference between the orthogonalization+aggressive and non-aggressive approaches is that,
  with orthogonalization,
  all the variance common across the accepted and rejected components is put in the accepted basket,
  and therefore it is not removed in the nuisance regression.
  In contrast, with the non-aggressive denoising,
  the shared variance is smartly split during the estimation process and therefore some amount is removed.

Now we can get started!
This walkthrough will show you how to perform different kinds of denoising on example data.

> For this walkthrough, we will use Python and the Nilearn package to denoised data.

Reviewer 2: ```suggestion
For this walkthrough, we will use Python and the Nilearn package to denoise data.
```

However, you can definitely accomplish all of these steps in other languages, like MATLAB,
or using other neuroimaging toolboxes, like AFNI.

> .. important::

Reviewer 2: I think you can remove this "Important" section. These are file names, for the same of the demo, there's no reason to say why these specific file names were used. The note on using different regressors is also indirectly mentioned elsewhere.

Author: Good catch. I'll remove it.

  If you do not want to include external regressors, you will need to adapt the code slightly;
  however, the necessary changes should be minimal.

Let's start by loading the necessary data.
No matter which type of denoising you want to use, you will need to include this step.

For this, you need the mixing matrix, the data you're denoising, a brain mask,
and an index of "bad" components.

.. code-block:: python

  import pandas as pd  # A library for working with tabular data

  # Files from fMRIPrep
  data_file = "sub-01_task-rest_space-MNI152NLin2009cAsym_desc-preproc_bold.nii.gz"
  mask_file = "sub-01_task-rest_desc-brain_mask.nii.gz"
  confounds_file = "sub-01_task-rest_desc-confounds_timeseries.tsv"

  # Files from tedana (after running on fMRIPrepped data)
  mixing_file = "sub-01_task-rest_desc-ICA_mixing.tsv"
  metrics_file = "sub-01_task-rest_desc-tedana_metrics.tsv"

  # Load the mixing matrix
  mixing_df = pd.read_table(mixing_file)  # Shape is time-by-components

  # Load the component table
  metrics_df = pd.read_table(metrics_file)
  rejected_columns = metrics_df.loc[metrics_df["classification"] == "rejected", "Component"]

>   accepted_columns = metrics_df.loc[metrics_df["classification"] == "accepted", "Component"]

Author: This is one thing that confuses me. Do we _want_ to orthogonalize our rejected components with respect to a bunch of "accepted" components we aren't actually confident about? We keep the (currently) ignored components in because we don't have anything better to do with them, given the amount of variance they explain or how unsure the decision tree is about whether they're BOLD or non-BOLD. I don't think that means I'm confident enough to say "these are definitely BOLD and I don't want my rejected components to share any variance with them".

>   accepted_columns = metrics_df.loc[metrics_df["classification"] == "accepted", "Component"]

Reviewer 2: The distinction between "accepted" and "ignored" is really fuzzy. In the current code, everything that isn't classified as rejected or ignored ends up in accepted. There are multiple reasons something might end up in ignored and I'm not sure I'd call all the actual criteria edge cases or things we are less confident about than the "accepted" components. This is one of several reasons the current terminology is problematic & "ignored" is being replaced with "accepted" plus more descriptive tags in the modularized code.

>   accepted_columns = metrics_df.loc[metrics_df["classification"] == "accepted", "Component"]

Author: I'm still wary of orthogonalizing rejected components on ones we're not sure about (or ones that don't explain a useful level of variance), but I also agree that the "ignored" label doesn't necessarily mean low-variance or ambiguous-classification, so I'll just leave as `== "accepted"`. Thanks!

##############################
Denoising Data with Components
##############################

Decomposition-based denoising methods like ``tedana`` will produce two important outputs: component time series and component classifications.

> The component classifications will indicate whether each componet is "good" (accepted) or "bad" (rejected).

Reviewer 2: ```suggestion
The component classifications will indicate whether each component is "good" (accepted) or "bad" (rejected).
```

To remove noise from your data, you can regress the "bad" components out of it, though there are multiple ways to accomplish this.

By default, ``tedana`` will perform a regression including both "good" and "bad" components, and then will selectively remove the "bad" components from the data.
This is colloquially known as "non-aggressive denoising".

However, users may wish to apply a different type of denoising, or to incorporate other regressors into their denoising step, and we will discuss these alternatives here.

.. admonition:: Independence vs. Orthogonality

>   ``tedana`` uses independent component analysis (ICA) to decompose the data into time series that are assumed to reflect meaningful underlying signals.
>   It then classifies each time series (a.k.a. component) as "accepted" (BOLD-like) or "rejected" (non-BOLD-like).
> 
>   The components from an ICA are `statistically independent`, rather than orthogonal.
>   Orthogonal time series are uncorrected, so there is no shared variance.
>   With independent components, however, the time series may be correlated with one another, and thus share variance.
>   If a rejected component shares meaningful variance with an accepted component,
>   then regressing the rejected components from your data may also remove meaningful signal associated with the accepted component as well.
>   This conundrum is what motivates the different approaches used for decomposition-based denoising methods described here.

Reviewer 2: ```suggestion
  ``tedana`` uses independent component analysis (ICA) to decompose the data into components which each have a spatial map of weights and time series that are assumed to reflect meaningful underlying signals.
  It then classifies each component) as "accepted" (BOLD-like) or "rejected" (non-BOLD-like). The data are denoised by taking the time series of the rejected components and regressing them from the data.

  ``tedana`` uses a spatial ICA, which means the components are `statistically independent` across space, not time. That means the time series from accepted and rejected components can share variance.
  If a rejected component shares meaningful variance with an accepted component,
  then regressing the rejected components' time series from your data may also remove meaningful signal associated with the accepted component as well.
  Depending on the application, people may take different approaches on how to handle variance that is shared between accepted and rejected components. These different approaches towards decomposition-based denoising methods are described here.
```
Maybe I shouldn't be reading this while tired, but I don't think this quite explains the issue. It's not about statistical independence vs orthogonality. It's that we're doing a spatial ICA that results in components that are  statistically independent across space. The distinction between orthogonality and statistical independence is not relevant to this specific issue.

Author: I've added your text in 6b3cb00. Thanks!
